### PR TITLE
It's just a flesh wound (HUD and health )

### LINKED ISSUE #222: Health
As a player, I want to have health, so that I am able to take some damage without instantly dying.

This task has the following subtasks:
- #223 Health - Implement Health Mechanic
- #224 Health - Show Health in HUD
- #225 Health - Create graphics for health in HUD

### LINKED ISSUE #223: Health - Implement Health Mechanic
Implement health mechanics into the game.

This is a subtask of #222 - Health

### LINKED ISSUE #224: Health - Show Health in HUD
As a player, I want to be able to see how much health I currently have, so that I can change my strategy accordingly.

This is a subtask of #222 - Health

### LINKED ISSUE #225: Health - Create graphics for health in HUD
The health displayed in the HUD should have proper graphics.

This is a subtask of #222 - Health

### LINKED ISSUE #226: HUD
As a player, I want to be able to see how much health, keys and bombs I currently have, so that I can manage my resources properly.

This task has the following subtasks:
- #227 HUD - Implement HUD
- #228 HUD - Link HUD with Inventory
- #229 HUD - Graphics for HUD

### LINKED ISSUE #227: HUD - Implement HUD
Implement the HUD in the game, so that it is properly displayed on the screen.

This is a subtask of #226 - HUD

### LINKED ISSUE #228: HUD - Link HUD with Inventory
Link the HUD with the player's inventory, so that it properly tracks the player's resources.

This is a subtask of #226 - HUD

### LINKED ISSUE #229: HUD - Graphics for HUD
Create graphics for Keys and Bombs, and display that in the HUD.

This is a subtask of #226 - HUD

### LINKED ISSUE #231: Cave looks and feel - Create Cave Textures
Create cave like textures for walls and the floor.

This is a subtask of #230 - Cave looks and feel

### PR DESCRIPTION
**Pull request by WytzeElhorst**
Adds a hud containing the healthbar, bombs and keys. Also adds health to the game and updates the textures and models for keys, bombs and doors

This closes #226 closes #227 closes #228 closes #229 (HUD) and closes #222 closes #223 closes #224 closes #225 (Health) and closes #231 (Cave Textures).

### CODE REVIEW

**Comment by ChielBruin** [comment]
You should remove the algoritmiek related files.

File: src/main/java/nl/tudelft/contextproject/controller/GameController.java [hunk 1]
```diff
@@ -109,7 +117,122 @@ public void onAction(String name, boolean isPressed, float tpf) {
...
+					keypicblue.setImage(Main.getInstance().getAssetManager(), "Textures/bluekeyicon.png", true);
+				} else {
+					keypicblue.setImage(Main.getInstance().getAssetManager(), "Textures/emptykeyicon.png", true);
+				}
+				if ((game.getPlayer().getInventory().containsColorKey(ColorRGBA.Red.set(1.0f, 0.0f, 0.0f, 0.0f)))) {
+					keypicred.setImage(Main.getInstance().getAssetManager(), "Textures/redkeyicon.png", true);
+				} else {
+					keypicred.setImage(Main.getInstance().getAssetManager(), "Textures/emptykeyicon.png", true);
+				}
+			}
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/controller/GameController.java]: We need to split up this method in smaller pieces, but we should do that in the next sprint.

File: src/main/java/nl/tudelft/contextproject/model/Inventory.java [hunk 2]
```diff
@@ -23,8 +23,23 @@ public Inventory() {
...
+		return keys;
+	}
+	/**
+	 * Returns the number of bombs in the inventory.
+	 * @return the number of bombs 
+	 */
+	public int numberOfBombs() {
+		return bombs;
 	}
-
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/Inventory.java:27]: Missing empty line.

File: src/main/java/nl/tudelft/contextproject/model/entities/Bomb.java [hunk 3]
```diff
@@ -49,28 +41,63 @@ public Spatial getSpatial() {
...
+			timer += tdf;
+			if (timer > 4) {
+				Explosion exp = new Explosion(40f);
+				Vector3f vec = this.getSpatial().getLocalTranslation();
+				exp.move(vec.x, vec.y, vec.z);
+				Main.getInstance().getCurrentGame().getEntities().add(exp);
+				this.setState(EntityState.DEAD);
+			}
 	}
+}
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/entities/Bomb.java]: Missing tabs, later in the file some more are missing.

File: src/main/java/nl/tudelft/contextproject/model/entities/Explosion.java [hunk 4]
```diff
@@ -53,11 +54,12 @@ public void update(float tpf) {
 			setState(EntityState.DEAD);
 			return;
 		}
-		if (collidesWithPlayer(scale.x / 2f)) {
-			//TODO Main.getInstance().getCurrentGame().getPlayer().damage(tpf);
-			System.out.println("BAM!");
+		if (collidesWithPlayer(scale.x / 5f)) {
+			player.takeDamage(tpf);
+			System.out.println(Main.getInstance().getCurrentGame().getPlayer().getHealth());
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/entities/Explosion.java]: Remove the Syso or replace with logger call

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java [hunk 5]
```diff
@@ -129,9 +138,8 @@ public void setSpatial(Spatial spatial) {
 
 	/**
 	 * Set character control.
-	 *
 	 * @param characterControl
-	 *		control to set
+	 * 				control to set
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java]: Remove some tab space.

File: src/main/java/nl/tudelft/contextproject/Main.java [hunk 6]
```diff
@@ -53,7 +56,7 @@
 	private Controller controller;
 	private WebServer webServer;
 	private List<TickListener> tickListeners = new LinkedList<>();
-
+	private BitmapFont guifont;
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/Main.java:61]: (nitpicking) new line after this line

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java [hunk 7]
```diff
@@ -167,35 +175,36 @@ public CharacterControl getPhysicsObject() {
 	public void onAction(String name, boolean isPressed, float tpf) {
```
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java]: This method is moved and can be removed here

**Comment by ChielBruin** [comment]
I made some comments about indentation and merging artifacts, after these are fixed i'm okay merging.

**Comment by ChielBruin** [comment]
Now it looks good! Next time try to create a separate pr for each addition.

File: src/main/java/nl/tudelft/contextproject/Main.java [hunk 8]
```diff
@@ -392,5 +404,21 @@ public boolean isControllerConnected() {
 		return sticks != null && sticks.length > 0;
 	}
 	
+	/**
+	 * Returns the BitmapFont.
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/Main.java:408]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/Main.java [hunk 9]
```diff
@@ -392,5 +404,21 @@ public boolean isControllerConnected() {
...
+	 */
+	public BitmapFont getGuiFont() {
+		return guifont;
+	}
+	
+
 	
+	/**
+	 * Returns the Controller().
+	 * @return the Controller()
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/Main.java]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/Inventory.java [hunk 10]
```diff
@@ -24,7 +24,21 @@ public Inventory() {
 		this.keys = 0;
 		this.bombs = 0;
 	}
-
+	
+	/**
+	 * Returns the number of keys in the inventory.
+	 * @return the number of keys
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/Inventory.java]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/Inventory.java [hunk 11]
```diff
@@ -24,7 +24,21 @@ public Inventory() {
...
+	 */
+	public int numberOfKeys() {
+		return keys;
+	}
+	/**
+	 * Returns the number of bombs in the inventory.
+	 * @return the number of bombs
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/Inventory.java]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/entities/Bomb.java [hunk 12]
```diff
@@ -69,8 +71,32 @@ public PhysicsControl getPhysicsObject() {
...
+	/**
+	 * activates the bomb, it will explode in 5 seconds.
+	 */
+	public void activate() {
+		this.active = true;
+	}
+
+	/**
+	 * Returns true if the bomb is active.
+	 * @return true if bomb is active.
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/Bomb.java]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/entities/Bomb.java [hunk 13]
```diff
@@ -69,8 +71,32 @@ public PhysicsControl getPhysicsObject() {
...
+	 */
+	public boolean getActive() {
+		return this.active;
+	}
+
+	/**
+	 * Returns the timer.
+	 * @return the timer
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/Bomb.java]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/entities/Door.java [hunk 14]
```diff
@@ -73,15 +66,15 @@ public PhysicsControl getPhysicsObject() {
...
 		if (rb == null) getPhysicsObject();
 
 		sp.move(x, y, z);
 		rb.setPhysicsLocation(rb.getPhysicsLocation().add(x, y, z));
 	}
-	
+
 	/**
 	 * Sets the color of the doors lock.
 	 * @param col
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/Door.java:81]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/entities/Door.java [hunk 15]
```diff
@@ -90,7 +83,7 @@ public void move(float x, float y, float z) {
 	public void setColor(ColorRGBA col) {
 		color = col;
 	}
-	
+
 	/**
 	 * Gets the color of the doors lock.
 	 * @return The color of the doors lock
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/Door.java]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java [hunk 16]
```diff
@@ -260,4 +264,33 @@ public Inventory getInventory() {
 	public void setInventory(Inventory inv) {
 		inventory = inv;
 	}
+	/**
+	 * Returns the player's health.
+	 * @return The player's health
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java [hunk 17]
```diff
@@ -260,4 +264,33 @@ public Inventory getInventory() {
...
+	 */
+	public float getHealth() {
+		return health;
+	}
+
+	/**
+	 * Sets a player's health.
+	 * @param heal
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java:282]: Comment in incorrect format.

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java [hunk 18]
```diff
@@ -260,4 +264,33 @@ public Inventory getInventory() {
...
+		if (heal > maxHealth) {
+			health = 3;
+		} else {
+			health = heal;
+		}
+	}
+
+	/**
+	 * Reduces a players health.
+	 * @param amount
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java:296]: Comment in incorrect format.

**Comment by Geweldig** [comment]
After my issues have been addressed I think this can be merged.

**Comment by Diocruel** [comment]
With the latest fixes I think it can be merged.

**Comment by Taeir** [comment]
As far as I'm concerned, this can be merged with the new changes.
Since I made some of those changes, it would be best if someone else merges this.

**Comment by Geweldig** [comment]
Looks good, merging.

**Pull request merged**